client/client.go

```diff
@@ -554,7 +557,37 @@ func (c *Client) do(req *http.Request, skipLoggingPayload bool) (*container.Cont
 			log.Printf("[DEBUG] HTTP response unique string %s %s %s", req.Method, req.URL.String(), bodyStr)
 		}
 
-		if (resp.StatusCode < 500 || resp.StatusCode > 504) && resp.StatusCode != 405 && !isApicNotFound(resp.StatusCode, bodyStr, resp.Header) {
+		// Handle session timeout for login based requests
+		reAuthError := false
+		if (resp.StatusCode == 403) {
+			obj, err := container.ParseJSON(bodyBytes)
+			if err != nil {
+				log.Printf("[DEBUG] Authorization error with status code 403")
+			} else {
+				errorMessage := obj.S("imdata").Index(0).S("error", "attributes", "text").String()
+				log.Printf("[DEBUG] Authorization error with status code 403: %+v", errorMessage)
+			}
+			log.Printf("[DEBUG] Checking max re-authentication retries: %v on %v", reAuthCounter, c.maxReAuthRetries)
+			if reAuthCounter < c.maxReAuthRetries {
+				log.Printf("[DEBUG] Retrying authentication after 403 error")
+				c.AuthToken = nil
+				req, err = c.InjectAuthenticationHeader(req, "")
+				if err == nil {
+					reAuthCounter = reAuthCounter + 1
+					// If re-authentication is successful, not counting previous query as an attempt
+					attempts = attempts - 1
+					log.Printf("[DEBUG] Retrying same request after status 403 and re-authentication")
+					continue
+				} else {
+					log.Printf("[ERROR] Error when retrying authentication after 403 error: %v", err)
+					reAuthError = true
+				}
+			} else {
+				log.Printf("[ERROR] Not retrying authentication after 403 error due to maximum re-authentication retries reached")
+			}
+		}
+
+		if (resp.StatusCode < 500 || resp.StatusCode > 504) && resp.StatusCode != 405 && !isApicNotFound(resp.StatusCode, bodyStr, resp.Header) && !reAuthError{
```

Comment:
I'm just clarifying:
shouldn't it be?
```
&& reAuthError{
```

Reply:
The if statement is a bit confusing. I had to check over it a few times.

It should use retry/backoff if there was a re-auth error which is the else condition. Hence, it should be `!reAuthError` here.

Reply:
@sajagana I think this would change the behaviour of the retry functions. By including 500 and 504 here it would end up skipping retry for 500, 504 which isn't the intention.
The client should retry when HTTP code is between 500-504, 405 or a re-auth error. The retry functionality is in the else condition.